ts/features/walletV3/payment/screens/WalletPaymentPickMethodScreen.tsx

```diff
@@ -44,51 +83,238 @@ const WalletPaymentPickMethodScreen = () => {
-      dispatch(walletPaymentPickPaymentMethod(wallet));
-    },
-    [dispatch]
-  );
+  const isLoading =
+    pot.isLoading(paymentMethodsPot) || pot.isLoading(userWalletsPots);
 
-  const handleContinue = () => {
-    navigation.navigate(WalletPaymentRoutes.WALLET_PAYMENT_MAIN, {
-      screen: WalletPaymentRoutes.WALLET_PAYMENT_PICK_PSP,
-      params: {
-        walletId: "123456",
-        paymentAmountInCents: 100
-      }
+  const paymentAmount = pot.getOrElse(paymentAmountPot, undefined);
+  const canContinue = selectedMethod !== undefined;
+
+  // ------------------------ HANDLERS --------------------------
+
+  const handleSelectSavedMethod = (walletId: string) => {
+    setSelectedMethod({
+      kind: "saved",
+      walletId
     });
   };
 
-  // TODO just for testing purposes
-  React.useEffect(() => {
-    if (pot.isSome(userWalletsPots) && !canContinue) {
-      const userWallets = userWalletsPots.value;
+  const handleSelectNotSavedMethod = (methodId: string) => {
+    setSelectedMethod({
+      kind: "generic",
+      methodId
+    });
+  };
 
-      if (userWallets.length > 0) {
-        handleMethodSelection(userWallets[0]);
-      }
+  const handleContinue = () => {
+    // todo:: should handle the case where the user
+    // selects a non saved method
+    if (paymentAmount && selectedMethod?.kind === "saved") {
+      pipe(
+        getSavedtMethodById(selectedMethod.walletId),
+        pot.toOption,
+        O.chainNullableK(
+          method => method && dispatch(walletPaymentPickPaymentMethod(method))
+        )
+      );
+      navigation.navigate(WalletPaymentRoutes.WALLET_PAYMENT_MAIN, {
+        screen: WalletPaymentRoutes.WALLET_PAYMENT_PICK_PSP
+      });
     }
-  }, [userWalletsPots, canContinue, handleMethodSelection]);
+  };
+
+  // -------------------------- LISTITEMS --------------------------
+
+  const savedMethodsListItems = useMemo(
+    () =>
+      isLoading
+        ? loadingRadios
+        : pipe(
+            userWalletsPots,
+            pot.toOption,
+            O.fold(
+              () => [],
+              // possibly void, so we coalesce to empty array
+              methods =>
+                (methods.map(mapSavedToRadioItem) ?? []) as Array<
+                  RadioItem<string>
+                >
+            )
+          ),
+    [isLoading, userWalletsPots]
+  );
+
+  const notSavedMethodsListItems = useMemo(
+    () =>
+      isLoading
+        ? loadingRadios
+        : pipe(
+            paymentMethodsPot,
+            pot.toOption,
+            O.fold(
+              () => [],
+              methods =>
+                methods.map(item => {
+                  const radio = mapGenericToRadioItem(item, paymentAmount);
+                  if (radio.disabled === true) {
+                    setShouldShowWarningBanner(true);
+                  }
+                  return radio;
+                })
+            )
+          ),
+    [isLoading, paymentMethodsPot, paymentAmount]
+  );
```

Comment:
I think the logic here is not correct and could be semplified as follows:

```suggestion
  const savedMethodsListItems = useMemo(
    () =>
      pipe(
        userWalletsPots,
        pot.toOption,
        O.map(methods => methods.map(mapSavedToRadioItem)),
        O.map(A.map(O.fromNullable)),
        O.map(A.compact),
        O.getOrElse(() => [] as Array<RadioItem<string>>)
      ),
    [userWalletsPots]
  );

  const genericMethodsListItems = useMemo(
    () =>
      pipe(
        paymentMethodsPot,
        pot.toOption,
        O.map(methods => methods.map(mapGenericToRadioItem)),
        O.getOrElse(() => [] as Array<RadioItem<string>>)
      ),
    [paymentMethodsPot]
```

The we should display the banner based on all methods, and not only for the generic ones:

```typescript
  React.useEffect(() => {
    if (!isLoading) {
      const hasDisabledMethods = [
        ...genericMethodsListItems,
        ...savedMethodsListItems
      ].find(item => item.disabled);
      setShouldShowWarningBanner(hasDisabledMethods !== undefined);
    }
  }, [isLoading, genericMethodsListItems, savedMethodsListItems]);
  ```

Reply:
while your pipe implementation indeed is cleaner, and even though we are working with relatively small lists, I think that running three more loops to find an entry we already are looping through is counterproductive.
It very well may be me just making a big deal out of a minimal performance loss; what do you think?

Reply:
> while your pipe implementation indeed is cleaner, and even though we are working with relatively small lists, I think that running three more loops to find an entry we already are looping through is counterproductive.

We are looping through a list of at most 20 items, it is practically nothing. instead we should pay attention to the number of renders.
The previous implementation sets the state every time it encounters a disabled method, the new implementation does it once.

> 
> Also, saved methods do not have a payment range as of writing this comment, so unless I missed some updates or there is another way to define a disabled method, I don't think it is possible.
> 
> It very well may be me just making a big deal out of a minimal performance loss; what do you think?

Saved method must be take into consideration as well. The range can be derived from the payment type but I don't think that currently we have this option. With this implementation we can easily add this feature once we have the information via BE.

Reply:
thinking about it, it is better to decouple the disabled and mapping logic; addressed in [6fd29bb](https://github.com/pagopa/io-app/pull/5350/commits/6fd29bb3a51428322dbf52470ab48c4a40e45ecb) 👍